Add idShop cache value for getLinkRewrite Category and Category CMS

Touxten opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Category link cache doesn't take shopId as a parameter
| Type?             | improvement
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | See issue
| Fixed issue or discussion?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/36794
| Sponsor company   | @friends-of-presta

clotairer committed:
Prevent Exception case #32748

dependabot[bot] committed:
build(deps-dev): bump jsdoc-to-markdown from 9.0.2 to 9.0.4 in /tests/UI

Bumps [jsdoc-to-markdown](https://github.com/jsdoc2md/jsdoc-to-markdown) from 9.0.2 to 9.0.4.
- [Release notes](https://github.com/jsdoc2md/jsdoc-to-markdown/releases)
- [Commits](https://github.com/jsdoc2md/jsdoc-to-markdown/compare/v9.0.2...v9.0.4)

---
updated-dependencies:
- dependency-name: jsdoc-to-markdown
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump @faker-js/faker from 9.0.3 to 9.1.0 in /tests/UI

Bumps [@faker-js/faker](https://github.com/faker-js/faker) from 9.0.3 to 9.1.0.
- [Release notes](https://github.com/faker-js/faker/releases)
- [Changelog](https://github.com/faker-js/faker/blob/next/CHANGELOG.md)
- [Commits](https://github.com/faker-js/faker/compare/v9.0.3...v9.1.0)

---
updated-dependencies:
- dependency-name: "@faker-js/faker"
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/eslint-plugin in /tests/UI

Bumps [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) from 8.8.1 to 8.12.2.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v8.12.2/packages/eslint-plugin)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump @faker-js/faker from 9.1.0 to 9.2.0 in /tests/UI

Bumps [@faker-js/faker](https://github.com/faker-js/faker) from 9.1.0 to 9.2.0.
- [Release notes](https://github.com/faker-js/faker/releases)
- [Changelog](https://github.com/faker-js/faker/blob/next/CHANGELOG.md)
- [Commits](https://github.com/faker-js/faker/compare/v9.1.0...v9.2.0)

---
updated-dependencies:
- dependency-name: "@faker-js/faker"
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump jsdoc-to-markdown from 9.0.4 to 9.0.5 in /tests/UI

Bumps [jsdoc-to-markdown](https://github.com/jsdoc2md/jsdoc-to-markdown) from 9.0.4 to 9.0.5.
- [Release notes](https://github.com/jsdoc2md/jsdoc-to-markdown/releases)
- [Commits](https://github.com/jsdoc2md/jsdoc-to-markdown/compare/v9.0.4...v9.0.5)

---
updated-dependencies:
- dependency-name: jsdoc-to-markdown
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump mocha from 10.7.3 to 10.8.2 in /tests/UI

Bumps [mocha](https://github.com/mochajs/mocha) from 10.7.3 to 10.8.2.
- [Release notes](https://github.com/mochajs/mocha/releases)
- [Changelog](https://github.com/mochajs/mocha/blob/main/CHANGELOG.md)
- [Commits](https://github.com/mochajs/mocha/compare/v10.7.3...v10.8.2)

---
updated-dependencies:
- dependency-name: mocha
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Progi1984 committed:
Fixed CI

ShaiMagal committed:
Update Module.php

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/eslint-plugin in /tests/UI

Bumps [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) from 8.12.2 to 8.13.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v8.13.0/packages/eslint-plugin)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump mysql2 from 3.11.3 to 3.11.4 in /tests/UI

Bumps [mysql2](https://github.com/sidorares/node-mysql2) from 3.11.3 to 3.11.4.
- [Release notes](https://github.com/sidorares/node-mysql2/releases)
- [Changelog](https://github.com/sidorares/node-mysql2/blob/master/Changelog.md)
- [Commits](https://github.com/sidorares/node-mysql2/compare/v3.11.3...v3.11.4)

---
updated-dependencies:
- dependency-name: mysql2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/parser in /tests/UI

Bumps [@typescript-eslint/parser](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/parser) from 8.13.0 to 8.14.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/parser/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v8.14.0/packages/parser)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/parser"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps-dev): bump @typescript-eslint/eslint-plugin in /tests/UI

Bumps [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) from 8.13.0 to 8.14.0.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v8.14.0/packages/eslint-plugin)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Hlavtox committed:
Unify error handling

Touxten committed:
Add idShop cache value for Link Rewrite Category and Catergory CMS

Touxten committed:
Improve and Add Eolia Suggestions

Eolia commented on classes/CMSCategory.php:
```diff
@@ -498,23 +498,37 @@ public static function checkBeforeMove($id_cms_category, $id_parent)
-		AND c.`id_cms_category` = ' . (int) $id_cms_category);
-        self::$_links[$id_cms_category . '-' . $id_lang] = $result['link_rewrite'];
+		WHERE `id_lang` = ' . (int) $idLang . '
+		AND c.`id_cms_category` = ' . (int) $idCMSCategory);
```
Hi Touxten, if you use $idShop in cache, use it in DB request ;)

        $result = Db::getInstance()->getRow('
            SELECT cl.`link_rewrite`
            FROM `'._DB_PREFIX_.'cms_category_lang` cl
            '.Shop::addSqlRestrictionOnLang('cl', (int)$idShop).'
            WHERE cl.`id_lang` = '.(int)$id_lang.'
            AND cl.`id_cms_category` = '.(int)$id_cms_category
        );

Touxten replied:
Yes actualy the PR is WIP because i have this issue :

Deprecated: Unparenthesized `a ?: b ? c : d` is deprecated. Use either `(a ?: b) ? c : d` or `a ?: (b ? c : d)` 

I need to ask in which branch the PR will go and also the minimum PHP version of the dev if it's 9.1

Eolia replied:
The use of nested ternaries is really not recommended, it is a source of problems and not very fluid.
I gave you a simple solution above ;)

Eolia replied:
And be careful because Validate::isUnsignedId() does not prevent zeros.

Eolia commented on classes/CMSCategory.php:
```diff
@@ -498,23 +498,37 @@ public static function checkBeforeMove($id_cms_category, $id_parent)
+        $idShop = $idShop ?: isset(Context::getContext()->shop->id)
+        ? (int) Context::getContext()->shop->id
+        : (int) Configuration::get('PS_SHOP_DEFAULT');
+
```
Ben...

        if(empty($idShop) || !Validate::isUnsignedId($idShop)) {
            $idShop = Context::getContext()->shop->id ?: Configuration::get('PS_SHOP_DEFAULT');
        }

dependabot[bot] committed:
build(deps-dev): bump @types/mocha from 10.0.9 to 10.0.10 in /tests/UI

Bumps [@types/mocha](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/mocha) from 10.0.9 to 10.0.10.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/mocha)

---
updated-dependencies:
- dependency-name: "@types/mocha"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Touxten commented:
New one : https://github.com/PrestaShop/PrestaShop/pull/37496
